### ISSUE TITLE
modify_dependencies does not recursively deserialize gufetokenizables

### ISSUE THREAD

**Issue by IAlibay** [OpenFreeEnergy/gufe#343]
## The use case

I have a subclass of a SolventComponent that has a SmallMoleculeComponent as one of its attributes.

## The problem

Roundtripping on keyed dict fails because (in part) `modify_dependencies` does not deserialize the SmallMoleculeComponent in the dict.

**Comment by dotsdl** [comment]
@IAlibay can you point us to where this class is implemented? Or provide it here as a snippet and show what it is you are trying that produces the failure?

**Comment by IAlibay** [comment]
@dotsdl of course, here is the code, including my current short time fix (had to reimplement the shallow/keyed dict method) https://github.com/IAlibay/pontibus/blob/develop/src%2Fpontibus%2Fcomponents%2Fextended_solvent_component.py

**Comment by dotsdl** [comment]
@atravitz this will require some initial investigation. From @IAlibay's description, this *shouldn't* be the behavior of `gufe`.

**Comment by IAlibay** [comment]
Just to update here, the issue seems specific to `keyed_dict` and `shallow_dict`.

So adding to @atravitz's example:

```
from gufe import SolventComponent
from extended_solvent_component import ExtendedSolventComponent


# round-trip a SolventComponent -- works
solvent = SolventComponent()
solvent_roundtrip = SolventComponent.from_dict(solvent.to_dict())
print(f"solvent roundtrip successful: {solvent == solvent_roundtrip}")


# round-trip an ExtendedSolventComponent -- works
ext_solvent = ExtendedSolventComponent()
ext_solvent_dict = ext_solvent.to_dict()
ext_solvent_roundtrip = ExtendedSolventComponent.from_dict(ext_solvent_dict)
print(f"extended solvent roundtrip successful: {ext_solvent == ext_solvent_roundtrip}")


# round-trip keyed dict -- fails
s = ExtendedSolventComponent.from_keyed_dict(ext_solvent.to_keyed_dict())

# round-trip shallow dict -- fails
s = ExtendedSolventComponent.from_shallow_dict(ext_solvent.to_shallow_dict())
```

**Comment by atravitz** [comment]
I was able to replicate this using @IAlibay's example here:https://github.com/OpenFreeEnergy/pontibus/blob/main/src/pontibus/components/extended_solvent_component.py

after looking into this with @dotsdl, it seems that removing lines 95-98

```
       if isinstance(self.solvent_molecule, SmallMoleculeComponent):
            solvent = self.solvent_molecule.to_dict()
        else:
            solvent = self.solvent_molecule
```
should fix things (along with removing `from_keyed_dict` and `from_shallow_dict`,  `from_dict_depth_one`, and `_from_dict_depth_one`.  Gufe's handling of keyed dicts and shallow dicts use the class's `to_dict()`, and so by explicitly casting `solvent_molecule` to a dict, gufe's handling of tokenizables was disrupted. 

I'm leaving this issue open until @IAlibay can confirm this is resolved for his use-case.

**Comment by dotsdl** [comment]
From @IAlibay, the proposed solution appears to be working well for him. Closing!
